### ISSUE TITLE
Awscr::S3::SignatureDoesNotMatch when repeating slashes

### ISSUE THREAD

**Issue by msa7** [taylorfinnell/awscr-s3#83]
I use DigitalOcean spaces and would like upload file into key like `2020/09/26/12/08/5b640175f6a4c62aba090ab1c5a69175`. Unfortunately I got exception: 

```
#<Awscr::S3::SignatureDoesNotMatch:>.
lib/awscr-s3/src/awscr-s3/http.cr:93:9 in 'handle_response!'
lib/awscr-s3/src/awscr-s3/http.cr:48:7 in 'put'
lib/awscr-s3/src/awscr-s3/client.cr:231:7 in 'put_object'
lib/awscr-s3/src/awscr-s3/file_uploader.cr:33:9 in 'upload'
src/console.cr:15:5 in 'run_console'
```

On other hand key `2020_09_26_12_08_5b640175f6a4c62aba090ab1c5a69175` works fine, but I need first one.

Is there any thing that could solve issue and any workaround?

**Comment by msa7** [comment]
Any news here?

As workaround I switched to run ruby process with gem aws-sdk-s3.

**Comment by brunto** [comment]
@msa7 could you share your code?

**Comment by taylorfinnell** [comment]
If you share a minimal snippet of what you're trying I can see if I can find the cause.

**Comment by msa7** [comment]
I just run ruby script for each file. Here is code.

Crystal part

```crystal
  def self.s3_upload_file(s3_key, file_path, origin_content_type, origin_file_name = nil)
    env = {
      "aws_access_key"        => Eda.config.digital_ocean_key,
      "aws_secret_access_key" => Eda.config.digital_ocean_secret,
      "s3_key"                => s3_key,
      "file_path"             => file_path,
      "origin_content_type"   => origin_content_type,
      "origin_file_name"      => origin_file_name,
    }

    Eda.ruby("upload_to_s3", env)
  end

  def self.ruby(name, env)
    file_name = "/app/src/service/ruby/#{name}.rb"
    stderr = IO::Memory.new
    stdout = IO::Memory.new
    status = Process.run(file_name, env: env, output: stdout, error: stderr, chdir: "/app/src/service/ruby")

    unless status.success?
      raise stderr.to_s
    end

    stdout.to_s
  end
```

Ruby part

```ruby
#!/usr/bin/env ruby

require "rubygems"
require "bundler/setup"
require "aws-sdk-s3"
require "json"

s3 = Aws::S3::Resource.new(
  access_key_id: ENV["aws_access_key"],
  secret_access_key: ENV["aws_secret_access_key"],
  endpoint: 'https://eda.fra1.digitaloceanspaces.com/',
  region: 'us-east-1'
)

options = {}
options[:content_type] = ENV["origin_content_type"] if ENV["origin_content_type"]
if ENV["origin_file_name"]
  options[:content_disposition] = "filename=\"#{ENV["origin_file_name"]}\""
  options[:metadata] = { name: ENV["origin_file_name"] }
end

obj = s3.bucket("eda_development").object(ENV["s3_key"])
obj.upload_file(ENV["file_path"], options)

puts true.to_json
```

**Comment by msa7** [comment]
And original broken  version was:

```crystal
    file_id = Random::Secure.hex
    dir_path = Time.utc.to_s("%Y/%m/%d")
    key = File.join(dir_path, "#{Time.utc.to_s("%H_%M")}_#{file_id}")

    Eda.s3_upload_file(key, file_path, origin_content_type, origin_file_name)


  # bug in Crystal lib https://github.com/taylorfinnell/awscr-s3/issues/83
  # not support file names with "/"
  def self.s3_upload(key, file_io, metadata = Hash(String, String).new)
    options = Awscr::S3::FileUploader::Options.new(with_content_types: false)
    uploader = Awscr::S3::FileUploader.new(Eda.s3, options)
    uploader.upload(Eda.s3_bucket, key, file_io, metadata)
  end
```

Issue happen then s3 key has `/`, see https://github.com/taylorfinnell/awscr-s3/issues/83#issue-710223728

**Comment by taylorfinnell** [comment]
Hi, this seems to work fine for me at least on S3. I can't test yet on DO.

```
require "awscr-s3"
AWS_ACCESS_KEY_ID="asdasd"
AWS_SECRET_ACCESS_KEY="asd"


CLIENT = Awscr::S3::Client.new(
  region: "us-east-1",
  aws_access_key: AWS_ACCESS_KEY_ID,
  aws_secret_key: AWS_SECRET_ACCESS_KEY
)

def s3_upload(key, file_io, metadata = Hash(String, String).new)
  options = Awscr::S3::FileUploader::Options.new(with_content_types: false)
  uploader = Awscr::S3::FileUploader.new(CLIENT, options)
  uploader.upload("bucket", key, file_io, metadata)
end

s3_upload("test/foo/bar/gif", File.open(File.expand_path("foo.gif")))
```

Were you attaching other headers like content type or anything in the original crystal failure?

**Comment by msa7** [comment]
Yep, I did setup a headers. The code was working perfectly for 1 year, but around 20 days ago I start receiving `Awscr::S3::SignatureDoesNotMatch`. I played around and found that changing s3_key from `Time.utc.to_s("%Y/%m/%d/%H/%M")` to `Time.utc.to_s("%Y_%m_%d_%H_%M")` makes code work again. But I would like to have `/` in s3_key, so I switched to ruby. Full version of the original code with tuned s3_key bellow, please see comments for broken s3_key.

```crystal
  def self.upload(io, origin_file_name, origin_content_type)
    file_id = Random::Secure.hex
    # file_path = Time.utc.to_s("%Y/%m/%d/%H/%M")
    # key = File.join(file_path, file_id)
    file_path = Time.utc.to_s("%Y_%m_%d_%H_%M") # workaround https://github.com/taylorfinnell/awscr-s3/issues/83
    key = "#{file_path}_#{file_id}"

    Dir.mkdir_p(File.join(UPLOAD_PATH, file_path))

    data = io.gets_to_end
    file_body = Base64.decode_string(data.split(",")[1])

    File.open(file_path(key), "w") do |file|
      file.puts(file_body)
    end

    file_matadata = {
      "x-amz-meta-name"     => origin_file_name,
      "Content-Type"        => origin_content_type,
      "Content-Disposition" => "filename=\"#{origin_file_name}\"",
    }
    Eda.log { "Uploading #{key} #{file_matadata}" }
    upload_to_s3(key, file_matadata)
end


  private def self.upload_to_s3(key, metadata)
    File.open(file_path(key), "r") do |file_io|
      Eda.s3_upload(key, file_io, metadata)
    end
  end

  def self.s3_upload(key, file_io, metadata = Hash(String, String).new)
    options = Awscr::S3::FileUploader::Options.new(with_content_types: false)
    uploader = Awscr::S3::FileUploader.new(Eda.s3, options)
    uploader.upload(Eda.s3_bucket, key, file_io, metadata)
  end
```

**Comment by taylorfinnell** [comment]
Thanks for the additional info, sadly I can't reproduce it against S3. It might be a DO issue?

**Comment by msa7** [comment]
Yep, ruby gem works, so look like DO compatible with S3. Might be the issue in the https://github.com/taylorfinnell/awscr-signer? I am worry about `The request URI can not contain repeating slashes`, could be a cause of the problem?

**Comment by andrius** [comment]
Odd, but it worked well in my case (AMS3 DO spaces)

**Comment by taylorfinnell** [comment]
I am pretty sure this is fixed after #87. Please re-open if it is not

**Comment by andrius** [comment]
Now I receive this with Crystal 1.0.0
